### PR TITLE
[d3d8] RefCount refactoring

### LINKED ISSUE #57: BloodRayne crashes on startup
The game crashes even before the intro videos are played. Luckily, I was able to get a trace this time.

### Software information
BloodRayne (Terminal Cut, but without using Ziggurat's d3d8.dll wrapper)

### System information
- GPU: Nvidia GTX 1080
- Driver: 515.65.01
- Wine version: 7.21 Staging
- DXVK version: 474d8c0

### Apitrace file(s)
[rayne1.trace.tar.xz](https://drive.google.com/file/d/1BuShmUQ7Bx78UOMwNJRBDcUlTXNGSbxl/view?usp=share_link)

### Log files
- d3d8.log: [rayne1_d3d8.log](https://github.com/AlpyneDreams/d8vk/files/10012874/rayne1_d3d8.log)
- d3d9.log: [rayne1_d3d9.log](https://github.com/AlpyneDreams/d8vk/files/10012875/rayne1_d3d9.log)

### LINKED ISSUE #78: BloodRayne 2 crashes on startup
The intro videos play, choppy as hell, and then the game crashes before getting to the main menu. I've captured a trace that doesn't reproduce the crash, unfortunately, but captures the problem with the intro videos (I'm hoping that's our main culprit here anyway).

### Software information
BloodRayne 2 (Terminal Cut, but without using Ziggurat's d3d8.dll wrapper)

### System information
- GPU: Intel Xe (TigerLake GT2)
- Driver: ANV/Mesa 23.0
- Wine version: 7.22 Staging
- DXVK version: d0f5903 (noticed first on 0.10)

### Apitrace file(s)
[rayne2.tar.xz](https://drive.google.com/file/d/19Iiu4XKuKTu27lyF23MRyEFhl4R3GHgR/view?usp=share_link)

### Log files
- d3d8.log: [rayne2_d3d8.log](https://github.com/AlpyneDreams/d8vk/files/10337937/rayne2_d3d8.log)
- d3d9.log: [rayne2_d3d9.log](https://github.com/AlpyneDreams/d8vk/files/10337942/rayne2_d3d9.log)

### LINKED ISSUE #90: Railroad Tycoon 3 - crash on graphic options change and game exit
The game crashes whenever one attempts to change the resolution or AA, and also on exit, with the following error:

error

Otherwise it renders everything correctly.

I've tried to capture a trace, however this is one of those games that force loads the d3d8.dll in system32, so I could only get a trace with d3d8to9 and WineD3D... and that combination, apparently, also has the same crashing problem.

WineD3D alone does not crash either on exit or on graphic options change.

### Software information
Railroad Tycoon 3 (GOG)

### System information
- GPU: Intel Xe (TigerLake GT2)
- Driver: ANV/Mesa 23.0.99
- Wine version: 7.22 Staging
- DXVK version: d0f5903

### Apitrace file(s)
[RT3.trace.tar.xz](https://drive.google.com/file/d/1MlKLE_stVG9cWiwH7JM1Oo0xIpzjGkNE/view?usp=share_link)

### Log files
- d3d8.log: [RT3_d3d8.log](https://github.com/AlpyneDreams/d8vk/files/10470413/RT3_d3d8.log)
- d3d9.log: [RT3_d3d9.log](https://github.com/AlpyneDreams/d8vk/files/10470414/RT3_d3d9.log)

### LINKED ISSUE #92: BlowOut crashes on startup
The game plays the intro cinematics only with sound (and a black screen) with WineD3D, and works just fine after that, but outright crashes on execution with d8vk.

P.S.: When playing back the trace, I get this warning in the d3d8.log file:

`warn:  D3D8 Application is windowed yet requested FullScreen_PresentationInterval 2147483648 (should be D3DPRESENT_INTERVAL_DEFAULT). This will be ignored.`

### Software information
BlowOut (GOG)

### System information
- GPU: RTX 3060
- Driver: 525.78.01
- Wine version: 7.22 Staging
- DXVK version: d0f5903

### Apitrace file(s)
[Blowout.trace.tar.xz](https://drive.google.com/file/d/1Di6T5jGEIgajfBQCy-UZgTkJchchkKPs/view?usp=share_link)

### Log files
- d3d8.log: [Blowout_d3d8.log](https://github.com/AlpyneDreams/d8vk/files/10499871/Blowout_d3d8.log)
- d3d9.log: [Blowout_d3d9.log](https://github.com/AlpyneDreams/d8vk/files/10499873/Blowout_d3d9.log)

### LINKED ISSUE #97: The hunt for Ref October - problems on shutdown/d3d device release
This is more of a meta-issue to keep track of the overall problem, so I'll break character and not follow the template for once.

There are already a number of open issues for games that crash on exit (e.g. BloodRayne, BloodRayne 2, Railroad Tycoon 3 etc.) with various errors, some outright pointing to dealocation/release problems in d8vk (see #90).

I'm also getting occasional crashes/hangs on exit with other d3d8 games. While it's hard to assess if they are caused by some other problem in Wine or d8vk, this may be the underlying cause behind them as well.

While retesting latest master, I've also noticed Duke Nukem - Manhattan Project errors out on exit, issuing the following messages from its own loggers:
```
*** ERROR *** <dx8> release refs (device)
*** ERROR *** <dx8> release refs 1 (d3d)
```
That doesn't cause any issue with the game or an outright crash, since in this particular case the game appears to handle them gracefully, however these errors do not show up with WineD3D.

P.S.: Should probably clarify that "The hunt for Ref October" is not a game, but a pun on a 90s movie & Tom Clancy novel. Since hunting for stray refs is bound to be as "thrilling", I thought it fitting.

### LINKED ISSUE #123: FireStarter - crash on resolution change
The game will crash when changing resolutions and reinitializing d3d. The problem is refcount related and doesn't occur with WineD3D.

image

## # Software information
FireStarter (GOG)

### System information
- GPU: RTX 3060
- Driver: 525.85.05
- Wine version: 8.3 Staging
- DXVK version: 8ddf18f

### Apitrace file(s)
N/A - game outright crashes when attempting to trace either with WineD3D or d8vk.

### Log files
- d3d8.log: [fsShell_d3d8.log](https://github.com/AlpyneDreams/d8vk/files/11026641/fsShell_d3d8.log)

### PR DESCRIPTION
**Pull request by WinterSnowfall**
Com objects already ref to the return pointer, so the current way of doing things will actually double refcount on certain calls, as can be seen in some traces. Doesn't fix any game per se as it stands and not sure if it will ultimately go anywhere, but some refcounts now appear to be more sane.

A good game to use for chasing the refs and comparing against how WineD3D refcounts is #92 . It in particular will crash due to an invalid ref count on an IDirect3DTexture8 object, but it also creates and releases a whole bunch of other things on startup before crashing which have invalid ref counts (some of which are addressed by this PR).

Next steps, based on the trace I'm looking at at least, would indicate:
- render target references need to be made private, however that blows up subresouces spectacularly, and m_container references no longer work for some reason (haven't figured it out yet)
- there is one more extra refcount on the parent IDirect3DTexture8 object which prevents its release, even if render targets are made private (this kinds of works if commenting out subresource refcounts to m_container in AddRef() and Release() ). I haven't yet found its annoyingly camouflaged hiding place.

Edit: Fixes #57, fixes #78, fixes #90, fixes #92, fixes #97, fixes #123.

### CODE REVIEW

**Comment by WinterSnowfall** [comment]
For the second iteration, I've managed to pull off making the render target refs private without blowing up everything (hopefully). I also found the extra ref, which I couldn't figure out before, it was m_textures in the device, which I've also switched to private.

In its current state the PR fixes #78 (which now works perfectly, without any crashes on exit or anything).

I haven't been able to get BloodRayne 1 and Blowout to work though. Both seem to have issues with setting the render target to the primary display for some reason. I'd like to figure that one out as well before I mark this as ready for review, but it may take some time, as I'm working on my C++-fu as well.

**Comment by WinterSnowfall** [comment]
Alright, I think I've reached the limit of what I can do/understand here. I believe this PR improves the ref/refcount situation to some degree. It fixes #78 #57 and #92. 

Sadly I could not fix the Railroad Tycoon 3 crash (#90) (which is also refcount related) or #123, although I know what would fix them, just not how to fix them without breaking other games. I'll comment about my findings on those issues.

P.S.: I've tried my best not to regress anything, but I guess anything is possible.

**Comment by WinterSnowfall** [comment]
I hope I'm at the final iteration... this properly fixes refcounting on textures/subresources, backbuffers + vertex & index buffers, but for some reason that entirely escapes me regresses Hitman Contracts, which will now crash on startup. I've tested over 50 other games and they're all fine.

The final piece of the puzzle was to not call back to d3d9 in SetRenderTarget if the render target or depth stencil addresses are unchanged, but rather just derp out and rely on what's cached for future fetch calls. This prevents crashing in #57 and #92. I have ported over some validations as well from d3d9 while I was at it, so we can more eagerly error out if we need to.

As to what it fixes, basically everything that was refcount related and mentioned above, namely: #57, #78, #90, #92 and #123.

Overall device count is still off compared to what I'm seeing in WineD3D traces, but games don't really seem to care about that too much, so meh. Might be worth looking into at some point though.

**Comment by WinterSnowfall** [comment]
I have solved the Hitman: Contracts regression with this last update.

The TLDR of it is the game constantly switches between two sets of render targets and depth stencils for some reason, so we need to cater for this case. The d3d8 spec also says with regards to SetRenderTarget:

`The previous depth-stencil surface's contents persist after a call to SetRenderTarget to disassociate the previous depth-stencil surface from the device.`

So we need to keep the previous depth stencil around, in case it gets re-associated. Apparently, Hitman also expects the same behavior for the render target, although this is not mentioned anywhere in the specs. In any case, the ref counting for the game in the d8vk trace now looks identical to what I'm seeing in WineD3D, so it's definitely doing the right thing.

I've retested the previously fixed titles as well and they are still fixed, as expected. I now hope absolutely nothing will regress because of this PR.

**Comment by WinterSnowfall** [comment]
Addressed another regression uncovered while testing Motor City Online. The game is just spuriously releasing stuff until it gets a 0 refcount (nice system, guys...). I've never seen any other games doing something this "innovative", but have added a sanity check to prevent unsigned underruns on subresource release, just to be safe. WineD3D seems to be handling it in a similar fashion.

**Comment by WinterSnowfall** [comment]
The refcount saga continues. While playing around a bit with the D3D8 SDK samples, I noticed some of them were erroring out while trying to go fullscreen, complaining about objects not being released properly. I've traced it down to overblown refcounts in relation to SetTexture() and stateblock Capture()s, and those bits are also fixed now. I'm not aware of any games being affected, but it's good to behave well here in any case.

**Comment by WinterSnowfall** [comment]
Updated again to fix a regression in Enclave (also present on main). The game depends on us derping out early in GetDepthStencilSurface in case of errors (added the equivalent for render targets for good measure).

**Comment by WinterSnowfall** [comment]
So, turns out I was being stupid (who would have guessed, right?) and abused SAFE_RELEASE, which also nulls the pointer that gets passed to it. State block derp.

This bit should also be fine now. Anyway, I've only seen this affect the "Water" D3D8 SDK sample, and possibly some of the other samples as well, but no actual games. All refcounts should now actually look identical to those in traces captured with WineD3D, including device refcounts, which is nice.

Even if arguably there could be room for improvement in some places, the refcounts are now correct, if not also in a very delicate balance (we really need to be careful not to add any additional public refs in the future). Would appreciate it if this PR gets merged in its current state and then later improved upon if needed, because it has been tested to hell and back by now.

Thanks to @Blisto91 for testing some of the heavy hitters that I don't own (think GTA SA & friends). I've also gone through most, if not all, of my 100+ d3d8 games collection and can confirm nothing regresses and that the above mentioned fixes actually work properly. 

Note that I've also included some non-related regression fixes (Enclave) in this PR. Perhaps not the best idea, but it sort of blended into the refactoring work really (and it's getting rather hard to keep track of all my other fixes filed as separate PRs).

File: src/d3d8/d3d8_device.h [hunk 1]
```diff
@@ -849,23 +897,25 @@ namespace dxvk {
...
-    INT                         m_baseVertexIndex = 0;
+    Com<D3D8IndexBuffer, false>        m_indices;
+    INT                                m_baseVertexIndex = 0;
 
     // TODO: Which of these should be a private ref
     std::vector<Com<D3D8Surface, false>> m_backBuffers;
     Com<D3D8Surface>            m_frontBuffer;
 
-    Com<D3D8Surface>            m_renderTarget;
+    Com<D3D8Surface, false>     m_renderTarget;
```
Comment (AlpyneDreams) [inline src/d3d8/d3d8_device.h:923]: I don't think these should all be private.
1. d3d8 docs say that `SetRenderTarget` increases ref on the new RT, but a private Com I think would not affect the public refcount?
2. Higher up you use `SAFE_RELEASE`, which calls `Release()`, after `m_renderTargetPrev.ref()` on these pointers, so you are increasing the private refcount, but decreasing the public refcount, which seems wrong. Ditto for `m_depthStencilPrev`

Similar deal for the depth-stencils, `SetIndices` and `m_indices`, `SetStreamSource` and `D3D8VBO`.

File: src/d3d8/d3d8_device.h [hunk 2]
```diff
@@ -368,30 +369,62 @@ namespace dxvk {
...
+          res = GetD3D9()->SetRenderTarget(0, surf->GetD3D9());
 
-        m_renderTarget = ref(surf);
+          if (FAILED(res)) return res;
+
+          // keep a temporary ref on the prev RT, since this could be a swap
+          D3D8Surface* pRenderTargetSwap = m_renderTargetPrev.ref();
+          m_renderTargetPrev = m_renderTarget;
+          m_renderTarget = surf;
+          SAFE_RELEASE(pRenderTargetSwap);
```
Comment (AlpyneDreams) [inline src/d3d8/d3d8_device.h]: See below wrt `m_renderTargetPrev`

File: src/d3d8/d3d8_device.h [hunk 3]
```diff
@@ -368,30 +369,62 @@ namespace dxvk {
...
+      if(likely(m_depthStencil.ptr() != zStencil)) {
+        res = GetD3D9()->SetDepthStencilSurface(D3D8Surface::GetD3D9Nullable(zStencil));
+
+        if (FAILED(res)) return res;
+
+        // keep a temporary ref on the prev DS, since this could be a swap
+        D3D8Surface* pDepthStencilSwap = m_depthStencilPrev.ref();
+        m_depthStencilPrev = m_depthStencil;
+        m_depthStencil = zStencil;
+        SAFE_RELEASE(pDepthStencilSwap);
```
Comment (AlpyneDreams) [inline src/d3d8/d3d8_device.h]: See below wrt `m_depthStencilPrev`

File: src/d3d8/d3d8_blit.cpp [hunk 4]
```diff
@@ -234,7 +234,7 @@ namespace dxvk {
         case D3DPOOL_SYSTEMMEM: {
 
           // RT (DEFAULT) -> SYSTEMMEM: Use GetRenderTargetData as fast path if possible
-          if ((srcDesc.Usage & D3DUSAGE_RENDERTARGET || m_renderTarget == src)) {
+          if ((srcDesc.Usage & D3DUSAGE_RENDERTARGET || m_renderTarget.ptr() == src.ptr())) {
```
Comment (AlpyneDreams) [inline src/d3d8/d3d8_blit.cpp:237]: Shouldn't be needed?

File: src/d3d8/d3d8_device_child.h [hunk 5]
```diff
@@ -32,6 +32,10 @@ namespace dxvk {
     }
     
     ULONG STDMETHODCALLTYPE Release() {
+      // ignore Release calls on objects with 0 refCount
+      if(unlikely(!this->m_refCount))
+        return this->m_refCount;
+
```
Comment (AlpyneDreams) [inline src/d3d8/d3d8_device_child.h:38]: This seems good. This behavior was supposed to come from `ComObjectClamp` before it was overridden here.

File: src/d3d8/d3d8_state_block.cpp
Comment (AlpyneDreams) [inline src/d3d8/d3d8_state_block.cpp:1]: Looks good.

File: src/d3d8/d3d8_texture.h [hunk 6]
```diff
@@ -103,7 +103,7 @@ namespace dxvk {
       return ptr;
     }
 
-    std::vector<Com<SubresourceType>> m_subresources;
+    std::vector<Com<SubresourceType, false>> m_subresources;
```
Comment (AlpyneDreams) [inline src/d3d8/d3d8_texture.h:106]: See above about private Com. This one might need to stay private, I'm not sure.

File: src/d3d8/d3d8_device.h [hunk 7]
```diff
@@ -723,7 +762,10 @@ namespace dxvk {
 
       D3D8VertexBuffer* buffer = static_cast<D3D8VertexBuffer*>(pStreamData);
 
-      m_streams[StreamNumber] = D3D8VBO {buffer, Stride};
+      const D3D8VBO& vbo = m_streams[StreamNumber];
+
+      if(likely(vbo.buffer.ptr() != buffer || vbo.stride != Stride))
+        m_streams[StreamNumber] = D3D8VBO {buffer, Stride};
```
Comment (AlpyneDreams) [inline src/d3d8/d3d8_device.h]: Why is this necessary?

File: src/d3d8/d3d8_device.h [hunk 8]
```diff
@@ -761,16 +803,20 @@ namespace dxvk {
 
       D3D8IndexBuffer* buffer = static_cast<D3D8IndexBuffer*>(pIndexData);
 
-      m_indices = buffer;
+      if(likely(m_indices.ptr() != buffer))
+        m_indices = buffer;
```
Comment (AlpyneDreams) [inline src/d3d8/d3d8_device.h]: Ditto.

File: src/d3d8/d3d8_device.h [hunk 9]
```diff
@@ -151,14 +151,15 @@ namespace dxvk {
...
       
       if (iBackBuffer >= m_backBuffers.size() || m_backBuffers[iBackBuffer] == nullptr) {
         Com<d3d9::IDirect3DSurface9> pSurface9;
         HRESULT res = GetD3D9()->GetBackBuffer(0, iBackBuffer, (d3d9::D3DBACKBUFFER_TYPE)Type, &pSurface9);
 
         if (FAILED(res)) return res;
         
-        m_backBuffers[iBackBuffer] = ref(new D3D8Surface(this, std::move(pSurface9)));
+        m_backBuffers[iBackBuffer] = new D3D8Surface(this, std::move(pSurface9));
         *ppBackBuffer = m_backBuffers[iBackBuffer].ref();
```
Comment (AlpyneDreams) [inline src/d3d8/d3d8_device.h:163]: Not 100% on this.

**Review by AlpyneDreams — CHANGES_REQUESTED** [review]
This PR has the right idea, and the fact that it supposedly fixes all those games is very good, but I am not certain we're doing things right here... See my review for more.

File: src/d3d8/d3d8_device.h (hunk 1 above)
Comment (WinterSnowfall) [reply src/d3d8/d3d8_device.h:923]: 1. The idea with private COMs in all these places is to have tighter control over the lifecycle of the cached objects and get to decide when they are released. The public refcount can still be adjusted on a private COM via .ref() (or ref(object) for that matter, they both do the same thing), while assignment and other operations will automagically handle the private refcount. Here is the COM object code I am referring to:

```
    T* ref() const {
      return dxvk::ref(m_ptr);
    }
```

```
  template<typename T>
  T* ref(T* object) {
    if (object != nullptr)
      object->AddRef();
    return object;
  }
```
Notice the latter is calling AddRef() directly, not incRef, so it is increasing the public refcount even on private COM objects.

Again, the main idea here is to keep the cached objects around as long as we need them, not according to what a game might decide by messing with the public refcount, while trying to match native public refcounts with privateComObject.ref(). As far as I can see from all the native traces I have compared against those captured with this PR, they now behave identically.

2. This bit is not ideal and I am open to suggestions on how to handle it better, however Hitman: Contracts seems to rely on this sort of behavior. What the game is doing on every frame is swapping between two render targets and depth stencils. It is not holding any public references to them, so their lifecycle is entirely controlled by our private refcounts. 

If I do not add a temporary ref here and release it after the swap is finalized, the game will crash. This is because the object that is about to be set is equal to the prev object, and the first operation, namely:

`m_depthStencilPrev = m_depthStencil;`

will 0 the private reference on m_depthStencilPrev, causing the object's destruction (which we don't want, because this is essentially equal to zStencil in case of a swap). So to keep the object around while I'm swapping it I am indeed increasing its public ref by 1 and releasing it when done.

I've used SAFE_RELEASE, because it also nulls the "swap" pointer after calling Release(), but simply calling Release in case of a not null pointer works as well. It's just that SAFE_RELEASE already has the needed logic in place.

File: src/d3d8/d3d8_device.h (hunk 2 above)
Comment (WinterSnowfall) [reply src/d3d8/d3d8_device.h]: Addressed, hopefully in point 2) above.

File: src/d3d8/d3d8_device.h (hunk 3 above)
Comment (WinterSnowfall) [reply src/d3d8/d3d8_device.h]: Addressed, hopefully in point 2) above.

File: src/d3d8/d3d8_blit.cpp (hunk 4 above)
Comment (WinterSnowfall) [reply src/d3d8/d3d8_blit.cpp:237]: One is now a private COM while the other is a public one so I get an error when I don't use .ptr() here.

File: src/d3d8/d3d8_device_child.h (hunk 5 above)
Comment (WinterSnowfall) [reply src/d3d8/d3d8_device_child.h:38]: This is only needed to prevent a regression with Motor City Online and this PR, since the game will call Release() even on objects with 0 refcount and won't stop calling Release() until it gets a 0 refcount again (going through uint32 values basically, as it underruns) . I haven't seen anything else rely on it tbh.

File: src/d3d8/d3d8_state_block.cpp
Comment (WinterSnowfall) [reply src/d3d8/d3d8_state_block.cpp:1]: :tada:

File: src/d3d8/d3d8_texture.h (hunk 6 above)
Comment (WinterSnowfall) [reply src/d3d8/d3d8_texture.h:106]: BloodRayne, BloodRayne2, Blowout, Railroad Tycoon 3, Firestarter and some of the D3D8 SDK apps seem to think otherwise. They do not behave correctly if we mess with public refcounts (especially on textures) beyond what is expected through d3d8 calls, and public COM objects also mess with them on assignment/release. So yes, this needs to stay private for sure.

File: src/d3d8/d3d8_device.h (hunk 7 above)
Comment (WinterSnowfall) [reply src/d3d8/d3d8_device.h]: It's not. I can remove it, but comparing is a bit faster then COM assignment shenanigans and I've seen d3d9 do the same. Nothing actually depends on it.

Edit: To elaborate, there are plenty of games that will do repeated calls with the same values at times.

File: src/d3d8/d3d8_device.h (hunk 8 above)
Comment (WinterSnowfall) [reply src/d3d8/d3d8_device.h]: Same as above.

File: src/d3d8/d3d8_device.h (hunk 9 above)
Comment (WinterSnowfall) [reply src/d3d8/d3d8_device.h:163]: As I've stated in point 1) above, privateComObject.ref() already increments the public ref on an object. We were doing it twice here.

**Comment by WinterSnowfall** [comment]
Hopefully the RT/DS swap logic is a bit more clear now, as it's gated to temporarily bump the public refcount only in case an actual swap is taking place. I've only seen Hitman: Contracts rely on this in practice, so other games shouldn't care much anyway.

Also removed the extra checks on SetIndeces/SetStreamSource which weren't crucial, as they weren't affecting refcounts.

File: src/d3d8/d3d8_device.h (hunk 1 above)
Comment (AlpyneDreams) [reply src/d3d8/d3d8_device.h:923]: Okay, I see, I always assumed ref called .incRef rather than ->AddRef. I am a little concerned about private refcounts purely because there may be cases where we forget to manually increase the refcount when a game expects a certain thing, but we forget to do it.

I think it could probably be solved another way, but if this increases compatability then it might be worth it.

File: src/d3d8/d3d8_texture.h (hunk 6 above)
Comment (AlpyneDreams) [reply src/d3d8/d3d8_texture.h:106]: You're right. I thought I made subresources ref publically to solve an incorrect refcount but I seem to recall it was actually to prevent unintended deallocation. This one is definitely good.

File: src/d3d8/d3d8_blit.cpp (hunk 4 above)
Comment (AlpyneDreams) [reply src/d3d8/d3d8_blit.cpp:237]: That's interesting, because Com<T, B> should have == implemented for any B
```cpp
    template<bool Public_>
    bool operator == (const Com<T, Public_>& other) const { return m_ptr == other.m_ptr; }
    template<bool Public_>
    bool operator != (const Com<T, Public_>& other) const { return m_ptr != other.m_ptr; }
```
But if the compiler is being dumb about it then I suppose leave it like that.

**Review by AlpyneDreams — APPROVED** [review]
Looks good now.
